odinmain/src/main/kotlin/me/odinmain/utils/skyblock/dungeon/DungeonUtils.kt

```diff
@@ -107,14 +118,31 @@ object DungeonUtils {
         var score = 0
         score += cryptCount.coerceAtMost(5)
         if (mimicKilled) score += 2
-        if (isPaul) score += 10
+        if ((isPaul && togglePaul == 0) || togglePaul == 2) score += 10
         return score
     }
 
+    inline val bloodDone: Boolean get() =
+        currentDungeon?.dungeonStats?.bloodDone ?: false
+
+    inline val score: Int get() {
+        val completed: Float = completedRoomCount.toFloat() + (if (!bloodDone) 1f else 0f) + (if (!inBoss) 1f else 0f)
+        val total: Float = if (totalRooms != 0) totalRooms.toFloat() else 36f
+
+        val exploration = floor((secretPercentage/floor.secretPercentage)/100f * 40f).coerceIn(0f, 40f).toInt() +
+                floor(completed/total * 60f).coerceIn(0f, 60f).toInt()
+
+        val skillRooms = floor(completed/total * 80f).coerceIn(0f, 80f).toInt()
+        val puzzlePenalty = puzzles.filter { it.status != PuzzleStatus.Completed }.size * 10
+        val skill = (20 + skillRooms - puzzlePenalty - (deathCount * 2 - 1).coerceAtLeast(0)).coerceIn(20, 100)
+
+        return exploration + skill + getBonusScore + 100
+    }
+
     inline val neededSecretsAmount: Int get() {
-        val deathModifier = deathCount * 2 - 1
+        val deathModifier = (deathCount * 2 - 1).coerceAtLeast(0)
         val scoreFactor = 40 - getBonusScore + deathModifier
-        return ceil(totalSecrets * scoreFactor / 40.0).toInt()
+        return ceil((totalSecrets * floor.secretPercentage) * scoreFactor / 40.0).toInt()
     }
 
```

Comment:
all of these things should realistically not be in this class but its fine for now someone will refactor later